File: python/wow_ironman_gen/wow_ironman_gen/toon.py
from dataclasses import dataclass
from wow_ironman_gen.races import Races
from wow_ironman_gen.classes import Classes
from wow_ironman_gen.factions import Factions
from wow_ironman_gen.genders import Genders


@dataclass
class Toon:
    """
    A WoW Toon with race, class, faction, and gender.
    """

    race: str
    wow_class: str
    faction: str
    gender: str

    @classmethod
    def random_toon(cls, data: dict):
        """
        Generates a random Toon, from the given data.
        :param data: dictionary with race keys and
                     permitted attributes as values
        :return: Toon or raises a ValueError
        """
        r = Races(data).random_race()
        if r is None:
            raise ValueError(f"invalid races: data={data}")
        c = Classes(data).random_class(r)
        if c is None:
            raise ValueError(f"invalid class: race={r}, data={data}")
        f = Factions(data).random_faction(r)
        if f is None:
            raise ValueError(f"invalid faction: race={r}, data={data}")
        g = Genders(data).random_gender(r)
        if g is None:
            raise ValueError(f"invalid gender: race={r}, data={data}")
        return Toon(race=r, wow_class=c, faction=f, gender=g)

    def __str__(self) -> str:
        return ' '.join([
            self.gender,
            self.faction,
            self.race,
            self.wow_class])

File: clojure/via_deps/wow-ironman-gen/src/wow_ironman_gen/toon.clj
(ns wow-ironman-gen.toon
  (:require [wow-ironman-gen.db :as db]
            [clojure.string :as str]))

(defrecord RandomToon [db race wow-class faction gender])

(defn new-random-toon [db]
  (let [race (db/random-race db)]
   (map->RandomToon {:db        db
                     :race      race
                     :wow-class (db/random-class db race)
                     :faction   (db/random-faction db race)
                     :gender    (db/random-gender db race)})))

(defn- pandaren?
  [{:keys [race] :as _toon}]
  (= race :pandaren))

(defn toon->str
  [{:keys [race wow-class faction gender]
    :as   toon}]
  (let [words (if (pandaren? toon)
                (map name [gender faction race wow-class])
                (map name [gender race wow-class]))]
    (str/join \space words)))
